pandas/core/series.py

```diff
@@ -2304,7 +2307,24 @@ def combine_first(self, other):
         other = other.reindex(new_index, copy=False)
         # TODO: do we need name?
         name = ops.get_op_result_name(self, other)  # noqa
-        rs_vals = com._where_compat(isna(this), other._values, this._values)
+        if not is_dtype_equal(this.dtype, other.dtype):
+            new_dtype = find_common_type([this.dtype, other.dtype])
+            if not is_dtype_equal(this.dtype, new_dtype):
+                this = this.astype(new_dtype)
+            if not is_dtype_equal(other.dtype, new_dtype):
+                other = other.astype(new_dtype)
+
+        if needs_i8_conversion(this.dtype):
```

Comment:
you can just use the in-built 

``other.where(isna(this), this)`` which already knows how to handle the dtypes

Reply:
then convert using ``pd.Index`` and use ``.where``

Reply:
really would prefer to fix #21546 first in any event.